tachyon/math/base/groups.h

```diff
@@ -76,25 +95,31 @@ class MultiplicativeGroup : public MultiplicativeSemigroup<G> {
     }
 
 #if defined(TACHYON_HAS_OPENMP)
-    using G2 = decltype(std::declval<G>().Inverse());
     size_t thread_nums = static_cast<size_t>(omp_get_max_threads());
     if (size >=
         size_t{1} << (thread_nums / kParallelBatchInverseDivisorThreshold)) {
       size_t chunk_size = base::GetNumElementsPerThread(groups);
       size_t num_chunks = (size + chunk_size - 1) / chunk_size;
+      std::atomic<bool> check_valid(true);
       OPENMP_PARALLEL_FOR(size_t i = 0; i < num_chunks; ++i) {
         size_t len = i == num_chunks - 1 ? size - i * chunk_size : chunk_size;
         absl::Span<const G> groups_chunk(std::data(groups) + i * chunk_size,
                                          len);
-        absl::Span<G2> inverses_chunk(std::data(*inverses) + i * chunk_size,
-                                      len);
-        DoBatchInverse(groups_chunk, inverses_chunk, coeff);
+        absl::Span<G> inverses_chunk(std::data(*inverses) + i * chunk_size,
+                                     len);
+        if (UNLIKELY(!DoBatchInverse(groups_chunk, inverses_chunk, coeff))) {
+          check_valid.store(false, std::memory_order_release);
+          continue;
+        }
+      }
+      if (UNLIKELY(!check_valid.load(std::memory_order_acquire))) {
```

Comment:
See https://stackoverflow.com/a/12448035
```suggestion
      if (UNLIKELY(!check_valid.load(std::memory_order_relaxed))) {
```